opte/src/engine/ip6.rs

```diff
@@ -262,113 +330,242 @@ impl From<&Ipv6Meta> for Ipv6Hdr {
+            // position.
+            next_header_pos = 0;
+
+            // Grab the remaining packet buffer, from the end of the previous
+            // header. This is where we'll start inserting the current extension
+            // header.
+            buf = &mut data[header_end..];
+
+            // Insert the bytes of each extension header, returning the number
+            // of octets written.
+            use IpProtocol::*;
+            let len = match extension {
+                HopByHop => {
+                    let hbh = hop_by_hop_header();
+                    let mut packet =
+                        Ipv6HopByHopHeader::new_checked(&mut buf).unwrap();
+                    hbh.emit(&mut packet);
+                    hbh.buffer_len()
+                }
+                Ipv6Frag => {
+                    let frag = fragment_header();
+                    let mut packet =
+                        Ipv6FragmentHeader::new_checked(&mut buf).unwrap();
+                    frag.emit(&mut packet);
+                    frag.buffer_len()
+                }
+                Ipv6Route => {
+                    let route = route_header();
+                    let mut packet =
+                        Ipv6RoutingHeader::new_checked(&mut buf).unwrap();
+                    route.emit(&mut packet);
+                    route.buffer_len()
+                }
+                Unknown(x) if x == &DDM_HEADER_ID => {
+                    // Starts with next_header, then a length excluding that.
+                    const DDM_HDR_LEN: usize = 15;
+                    buf[1] = DDM_HDR_LEN as u8;
+                    DDM_HDR_LEN + 1
+                }
+                _ => unimplemented!(
+                    "Extension header {:#?} unsupported",
+                    extension
+                ),
+            };
+
+            // Move the position markers to the new header.
+            header_start = header_end;
+            header_end += len;
         }
+
+        // Set the last header to point to the ULP
+        data[header_start] = u8::from(IpProtocol::Tcp);
+
+        (data, header_end)
     }
-}
 
-#[cfg(test)]
-mod test {
-    fn from_pairs() {
-        let ip6 = super::Ipv6Addr::from([
-            0x2601, 0x0284, 0x4100, 0xE240, 0x0000, 0x0000, 0xC0A8, 0x01F5,
-        ]);
+    // Test every permuation of the supported extension headers, verifying the
+    // computed lengths of:
+    //
+    // - Payload length
+    // - ULP length
+    // - Extension header length
+    // - Full header length
+    #[test]
+    fn test_extension_header_lengths_ok() {
+        for n_extensions in 0..SUPPORTED_EXTENSIONS.len() {
+            for extensions in
+                SUPPORTED_EXTENSIONS.into_iter().permutations(n_extensions)
+            {
+                let (buf, pos) = generate_test_packet(extensions.as_slice());
+                let mut simple_packet = SimplePacket::from(buf);
```

Comment:
Why do we need `SimplePacket`? Can't we just use something like `PacketReader::new(Packet<Initialized>::copy(buf), ())`?

Reply:
It sure likes like it. See, the thing is, you have to _know that's there_. 😬 Thanks for pointing it out, I'll switch everything over.